src/fu-console.c

```diff
@@ -0,0 +1,753 @@
+fu_console_spin_cb(gpointer user_data)
+{
+	FuConsole *self = FU_CONSOLE(user_data);
+
+	/* move the spinner index up to down */
+	fu_console_spin_inc(self);
+
+	/* update the terminal */
+	fu_console_refresh(self);
+
+	return G_SOURCE_CONTINUE;
+}
+
+static void
+fu_console_spin_end(FuConsole *self)
+{
+	if (self->timer_source != NULL) {
+		g_source_destroy(self->timer_source);
+		self->timer_source = NULL;
+
+		/* reset when the spinner has been stopped */
+		g_timer_start(self->time_elapsed);
+	}
+
+	/* go back to the start when we next go into unknown percentage mode */
+	self->spinner_idx = 0;
+	self->spinner_count_up = TRUE;
+}
+
+static void
+fu_console_spin_start(FuConsole *self)
+{
+	if (self->timer_source != NULL)
+		g_source_destroy(self->timer_source);
+	self->timer_source = g_timeout_source_new(40);
+	g_source_set_callback(self->timer_source, fu_console_spin_cb, self, NULL);
+	g_source_attach(self->timer_source, self->main_ctx);
+}
+
+/**
+ * fu_console_set_progress:
+ * @self: A #FuConsole
+ * @status: A #FwupdStatus
+ * @percentage: unsigned integer
+ *
+ * Refreshes a console
+ *
+ * Since: 0.9.7
+ **/
+void
+fu_console_set_progress(FuConsole *self, FwupdStatus status, guint percentage)
+{
+	g_return_if_fail(FU_IS_CONSOLE(self));
+
+	/* not useful */
+	if (status == FWUPD_STATUS_UNKNOWN)
+		return;
+
+	/* ignore duplicates */
+	if (self->status == status && self->percentage == percentage)
+		return;
+
+	/* cache */
+	self->status = status;
+	self->percentage = percentage;
+
+	/* dumb */
+	if (!self->interactive) {
+		g_print("%s: %u%%\n", fu_console_status_to_string(status), percentage);
+		return;
+	}
+
+	/* if the main loop isn't spinning and we've not had a chance to
+	 * execute the callback just do the refresh now manually */
+	if (percentage == 0 && status != FWUPD_STATUS_IDLE &&
+	    self->status != FWUPD_STATUS_UNKNOWN) {
+		if ((g_get_monotonic_time() - self->last_animated) / 1000 > 40) {
+			fu_console_spin_inc(self);
+			fu_console_refresh(self);
+		}
+	}
+
+	/* enable or disable the spinner timeout */
+	if (percentage > 0) {
+		fu_console_spin_end(self);
+	} else {
+		fu_console_spin_start(self);
+	}
+
+	/* update the terminal */
+	fu_console_refresh(self);
+}
+
+/**
+ * fu_console_set_interactive:
+ * @self: A #FuConsole
+ * @interactive: #gboolean
+ *
+ * Marks the console as interactive or not
+ **/
+void
+fu_console_set_interactive(FuConsole *self, gboolean interactive)
+{
+	g_return_if_fail(FU_IS_CONSOLE(self));
+	self->interactive = interactive;
+}
+
+/**
+ * fu_console_set_status_length:
+ * @self: A #FuConsole
+ * @len: unsigned integer
+ *
+ * Sets the length of the console status
+ **/
+void
+fu_console_set_status_length(FuConsole *self, guint len)
+{
+	g_return_if_fail(FU_IS_CONSOLE(self));
+	g_return_if_fail(len > 3);
+	self->length_status = len;
+}
+
+/**
+ * fu_console_set_percentage_length:
+ * @self: A #FuConsole
+ * @len: unsigned integer
+ *
+ * Sets the length of the progressba percentage
```

Comment:
progressbar
also explain why limit is less than 3

Reply:
done, and done.